### PR TITLE
ra-tls: enable local report attestation function

### PR DESCRIPTION
**Pull request by YangLiang3**

### CODE REVIEW

**Comment by jiazhang0** [comment]
Please add the fix line for #532. @YangLiang3

File: ra-tls/Makefile [hunk 1]
```diff
@@ -53,14 +57,15 @@ $(BINDIR)/Wolfssl_Enclave.signed.so: $(BINDIR)
 	$(MAKE) -C stub-enclave && \
 	  cp -f stub-enclave/Wolfssl_Enclave.signed.so "$@"
 
+
```
Comment (jiazhang0) [inline ra-tls/Makefile:60]: Plz remove this empty line.

File: ra-tls/sgx-ra-tls/wolfssl-la-attester.c [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+{
+	la_wolfssl_create_key_and_x509(der_key, der_key_len,
+			der_cert, der_cert_len);
+}
+
+int enc_la_sgx_verify_report(sgx_report_t* report)
+{
+	sgx_report_t report_t;
+	memcpy(&report_t, report, sizeof(sgx_report_t));
+        sgx_status_t status = sgx_verify_report(&report_t);
```
Comment (jiazhang0) [inline ra-tls/sgx-ra-tls/wolfssl-la-attester.c:106]: Plz fix this indent issue.

File: ra-tls/sgx-ra-tls/ra-challenger.c [hunk 3]
```diff
@@ -127,7 +138,7 @@ int extract_x509_extension
     if (rc == -1) return -1;
     
     assert(ext_data != NULL);
-    assert(ext_data_len <= data_max_len);
+//    assert(ext_data_len <= data_max_len);
```
Comment (jiazhang0) [inline ra-tls/sgx-ra-tls/ra-challenger.c:141]: Why comment this line?

File: ra-tls/pal/Makefile [hunk 4]
```diff
@@ -61,7 +62,7 @@ endif
 
 #App_C_Files := App.c server-tls.c sgxsdk-ra-attester_u.c ias-ra.c
 #App_C_Files := App.c sgxsdk-ra-attester_u.c
-App_C_Fils := App.c
+App_C_Files := App.c
```
Comment (jiazhang0) [inline ra-tls/pal/Makefile:65]: Why rename to App_C_Files?

File: ra-tls/elv/Makefile [hunk 5]
```diff
@@ -3,31 +3,46 @@ TOPDIR := $(shell readlink -f ..)
...
 endif
 
-CFLAGS += -std=gnu99 -I$(SGX_RA_TLS) -I$(SGX_SDK)/include -I$(INCDIR) $(SGX_DCAP_INC) -fPIC
-#CFLAGSERRORS := -Wall -Wextra -Wwrite-strings -Wlogical-op -Wshadow -Werror
-CFLAGS += $(CFLAGSERRORS) -g -O0 -DWOLFSSL_SGX_ATTESTATION -DWOLFSSL_CERT_EXT # -DDEBUG -DDYNAMIC_RSA
-CFLAGS += -DSGX_GROUP_OUT_OF_DATE
+Wolfssl_C_Extra_Flags := -DWOLFSSL_SGX -DUSE_WOLFSSL
+Wolfssl_Include_Paths := -I$(WOLFSSL_ROOT) \
+	-I$(WOLFSSL_ROOT)/wolfcrypt
+App_Include_Paths := $(Wolfssl_Include_Paths) -I$(SGX_SDK)/include -I$(SGX_RA_TLS_ROOT) -I$(INCDIR) $(SGX_DCAP_INC) -I$(shell readlink -f .)
```
Comment (jiazhang0) [inline ra-tls/elv/Makefile:31]: The definition of SGX_SDK is remove?

File: ra-tls/elv/Makefile [hunk 6]
```diff
@@ -3,31 +3,46 @@ TOPDIR := $(shell readlink -f ..)
...
+       SGX_COMMON_CFLAGS += -O2
 endif
 
-CFLAGS += -std=gnu99 -I$(SGX_RA_TLS) -I$(SGX_SDK)/include -I$(INCDIR) $(SGX_DCAP_INC) -fPIC
-#CFLAGSERRORS := -Wall -Wextra -Wwrite-strings -Wlogical-op -Wshadow -Werror
-CFLAGS += $(CFLAGSERRORS) -g -O0 -DWOLFSSL_SGX_ATTESTATION -DWOLFSSL_CERT_EXT # -DDEBUG -DDYNAMIC_RSA
-CFLAGS += -DSGX_GROUP_OUT_OF_DATE
+Wolfssl_C_Extra_Flags := -DWOLFSSL_SGX -DUSE_WOLFSSL
+Wolfssl_Include_Paths := -I$(WOLFSSL_ROOT) \
+	-I$(WOLFSSL_ROOT)/wolfcrypt
```
Comment (jiazhang0) [inline ra-tls/elv/Makefile:30]: Combine these 2 lines into a single line.

File: ra-tls/elv/Makefile [hunk 7]
```diff
@@ -3,31 +3,46 @@ TOPDIR := $(shell readlink -f ..)
...
+App_Include_Paths := $(Wolfssl_Include_Paths) -I$(SGX_SDK)/include -I$(SGX_RA_TLS_ROOT) -I$(INCDIR) $(SGX_DCAP_INC) -I$(shell readlink -f .)
+App_C_Flags := $(SGX_COMMON_CFLAGS) -fPIC -Wno-attributes -Wall -Wno-unused-const-variable $(App_Include_Paths) $(Wolfssl_C_Extra_Flags)
+
 ifdef ECDSA
-  CFLAGS += -DRATLS_ECDSA
+App_C_Flags += -DRATLS_ECDSA
```
Comment (jiazhang0) [inline ra-tls/elv/Makefile:35]: Plz add an indent.

File: ra-tls/elv/Makefile [hunk 8]
```diff
@@ -3,31 +3,46 @@ TOPDIR := $(shell readlink -f ..)
...
 endif
 
-CC ?= gcc
-GO ?= go
+ifdef LA
+App_C_Flags += -DLA_REPORT
```
Comment (jiazhang0) [inline ra-tls/elv/Makefile:39]: ditto

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: ra-tls/sgx-ra-tls/ra-challenger.c (hunk 3 above)
Comment (YangLiang3) [reply ra-tls/sgx-ra-tls/ra-challenger.c:141]: will recover it

File: ra-tls/pal/Makefile (hunk 4 above)
Comment (YangLiang3) [reply ra-tls/pal/Makefile:65]: it should be typo. App_C_Files not App_C_Fils

File: ra-tls/elv/Makefile (hunk 5 above)
Comment (YangLiang3) [reply ra-tls/elv/Makefile:31]: SGX_SDK is exported in top Makefile

File: ra-tls/elv/Makefile (hunk 7 above)
Comment (YangLiang3) [reply ra-tls/elv/Makefile:35]: Thanks

File: ra-tls/patch/wolfssl.patch [hunk 9]
```diff
@@ -108,7 +108,31 @@ index 12ada17af..43cafec21 100644
...
  #include <wolfssl/wolfcrypt/random.h>
  #include <wolfssl/wolfcrypt/hash.h>
-@@ -11448,7 +11449,10 @@ int wc_PemPubKeyToDer(const char* fileName,
+@@ -167,6 +168,23 @@ int tsip_tls_CertVerify(const byte *cert, word32 certSz,
+                         word32 key_e_start, word32 key_e_len,
+                         byte *tsip_encRsaKeyIdx);
+ #endif
++
++#ifdef LA_REPORT
++#ifdef WOLFSSL_SGX
```
Comment (jiazhang0) [inline ra-tls/patch/wolfssl.patch:117]: Change to
```C
#if defined(LA_REPORT) && defined(WOLFSSL_SGX)
```

File: ra-tls/sgx-ra-tls/wolfssl-ra-challenger.c [hunk 10]
```diff
@@ -429,13 +430,17 @@ int verify_sgx_cert_extensions
     uint32_t der_crt_len
 )
 {
+
```
Comment (jiazhang0) [inline ra-tls/sgx-ra-tls/wolfssl-ra-challenger.c:433]: Plz remove this empty line.

File: ra-tls/sgx-ra-tls/wolfssl-ra-attester.c [hunk 11]
```diff
@@ -135,8 +135,10 @@ void ecdsa_create_key_and_x509
 	ecdsa_wolfssl_create_key_and_x509(der_key, der_key_len,
 			der_cert, der_cert_len);
 }
-#else
+#endif
 
+#ifndef RATLS_ECDSA
+#ifndef LA_REPORT
```
Comment (jiazhang0) [inline ra-tls/sgx-ra-tls/wolfssl-ra-attester.c:141]: Change to 
```C
if !defined(RATLS_ECDSA) && !defined(LA_REPORT)
```

File: ra-tls/sgx-ra-tls/wolfssl-la-challenger.c [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+	la_get_report_from_cert(der_crt, der_crt_len, &report);
+	enc_la_sgx_verify_report(geid, &retval, &report);
+	if (retval == 0)
+	{
+		printf("Verify local report successfully!\n");
+		return 0;
+	} else {
+		printf("Verify local report failure!\n");
+		return -1;
+	}
```
Comment (jiazhang0) [inline ra-tls/sgx-ra-tls/wolfssl-la-challenger.c:51]: Change to:
```C
if (retval == 0) {
   ...
}

printf("Verify local report failure!\n");
return -1;
```

File: ra-tls/sgx-ra-tls/wolfssl-la-challenger.c [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+#include <wolfssl/ssl.h>
+#include <wolfssl/wolfcrypt/asn.h>
+#include <wolfssl/wolfcrypt/asn_public.h>
+#include <wolfssl/wolfcrypt/coding.h>
+#include <wolfssl/wolfcrypt/rsa.h>
+#include <wolfssl/wolfcrypt/sha256.h>
+#include <wolfssl/wolfcrypt/signature.h>
+#include <sgx_urts.h>
+
+extern sgx_enclave_id_t geid;
```
Comment (jiazhang0) [inline ra-tls/sgx-ra-tls/wolfssl-la-challenger.c:20]: Use the name `g_eid`.

File: ra-tls/sgx-ra-tls/wolfssl-la-attester.c [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    strncpy(crt.subject.locality, "Hillsboro", CTC_NAME_SIZE);
+    strncpy(crt.subject.org, "Intel Inc.", CTC_NAME_SIZE);
+    strncpy(crt.subject.unit, "Intel Labs", CTC_NAME_SIZE);
+    strncpy(crt.subject.commonName, "SGX rocks!", CTC_NAME_SIZE);
+    strncpy(crt.subject.email, "webmaster@intel.com", CTC_NAME_SIZE);
+
+    memcpy(crt.lareport, report, sizeof(sgx_report_t));
+    crt.lareportSz = sizeof(sgx_report_t);
+
+    RNG    rng;
```
Comment (jiazhang0) [inline ra-tls/sgx-ra-tls/wolfssl-la-attester.c:53]: Change to
```C
RNG rng;
```

File: ra-tls/sgx-ra-tls/sgxsdk-ra-attester_t.c [hunk 15]
```diff
@@ -11,6 +11,7 @@
 /* Trusted portion (called from within the enclave) to do remote
    attestation with the SGX SDK.  */
 #ifndef RATLS_ECDSA
+#ifndef LA_REPORT
```
Comment (jiazhang0) [inline ra-tls/sgx-ra-tls/sgxsdk-ra-attester_t.c:14]: Change to
```C
#if !defined(RATLS_ECDSA) && !defined(LA_REPORT)
```

**Review by jiazhang0 — CHANGES_REQUESTED** [review]
There is no big issue. Just code style cleanup.

File: ra-tls/sgx-ra-tls/wolfssl-ra-challenger.c (hunk 10 above)
Comment (YangLiang3) [reply ra-tls/sgx-ra-tls/wolfssl-ra-challenger.c:433]: ok, thanks

File: ra-tls/sgx-ra-tls/wolfssl-ra-attester.c (hunk 11 above)
Comment (YangLiang3) [reply ra-tls/sgx-ra-tls/wolfssl-ra-attester.c:141]: ok, thanks

File: ra-tls/sgx-ra-tls/wolfssl-la-challenger.c (hunk 13 above)
Comment (YangLiang3) [reply ra-tls/sgx-ra-tls/wolfssl-la-challenger.c:20]: ok, thanks

File: ra-tls/sgx-ra-tls/sgxsdk-ra-attester_t.c (hunk 15 above)
Comment (YangLiang3) [reply ra-tls/sgx-ra-tls/sgxsdk-ra-attester_t.c]: done

File: ra-tls/sgx-ra-tls/wolfssl-la-attester.c (hunk 14 above)
Comment (YangLiang3) [reply ra-tls/sgx-ra-tls/wolfssl-la-attester.c]: done

File: ra-tls/sgx-ra-tls/wolfssl-la-challenger.c (hunk 12 above)
Comment (YangLiang3) [reply ra-tls/sgx-ra-tls/wolfssl-la-challenger.c:47]: done

**Pull request merged**